From 51854aa7e60f9e05bda2226d3a63e57f2e2b8d08 Mon Sep 17 00:00:00 2001
From: David Milburn <dmilburn@redhat.com>
Date: Fri, 25 Apr 2014 15:27:02 -0400
Subject: [ata] ahci: Make distinct names for ports in /proc/interrupts

Message-id: <1398439625-8238-2-git-send-email-dmilburn@redhat.com>
Patchwork-id: 79180
O-Subject: [RHEL7.1 PATCH 1/4] AHCI: Make distinct names for ports in /proc/interrupts
Bugzilla: 1067838
RH-Acked-by: Jes Sorensen <Jes.Sorensen@redhat.com>
RH-Acked-by: Mark Langsdorf <mlangsdo@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>

commit b29900e62598cecd519c9ab2b8e4d03f8ebf702d
Author: Alexander Gordeev <agordeev@redhat.com>
Date:   Wed May 22 08:53:48 2013 +0900

    AHCI: Make distinct names for ports in /proc/interrupts

    Currently all interrupts assigned to AHCI ports show up in
    '/proc/interrupts' as 'ahci'. This fix adds port numbers as
    suffixes and hence makes the descriptions distinct.

    Reported-by: Jan Beulich <JBeulich@suse.com>
    Signed-off-by: Alexander Gordeev <agordeev@redhat.com>
    Signed-off-by: Tejun Heo <tj@kernel.org>

https://bugzilla.redhat.com/show_bug.cgi?id=1067838
https://brewweb.devel.redhat.com/taskinfo?taskID=7368411
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/ata/ahci.c b/drivers/ata/ahci.c
index 3d67f76..bfa8db6 100644
--- a/drivers/ata/ahci.c
+++ b/drivers/ata/ahci.c
@@ -1152,9 +1152,11 @@ int ahci_host_activate(struct ata_host *host, int irq, unsigned int n_msis)
   return rc;
 
  for (i = 0; i < host->n_ports; i++) {
+  struct ahci_port_priv *pp = host->ports[i]->private_data;
+
   rc = devm_request_threaded_irq(host->dev,
    irq + i, ahci_hw_interrupt, ahci_thread_fn, IRQF_SHARED,
-   dev_driver_string(host->dev), host->ports[i]);
+   pp->irq_desc, host->ports[i]);
   if (rc)
    goto out_free_irqs;
  }
diff --git a/drivers/ata/ahci.h b/drivers/ata/ahci.h
index 10b14d4..d71a842 100644
--- a/drivers/ata/ahci.h
+++ b/drivers/ata/ahci.h
@@ -306,6 +306,7 @@ struct ahci_port_priv {
  int   fbs_last_dev; /* save FBS.DEV of last FIS */
  /* enclosure management info per PM slot */
  struct ahci_em_priv em_priv[EM_MAX_SLOTS];
+ char   *irq_desc; /* desc in /proc/interrupts */
 };
 
 struct ahci_host_priv {
diff --git a/drivers/ata/libahci.c b/drivers/ata/libahci.c
index 020df80..3283b71 100644
--- a/drivers/ata/libahci.c
+++ b/drivers/ata/libahci.c
@@ -2257,6 +2257,16 @@ static int ahci_port_start(struct ata_port *ap)
  if (!pp)
   return -ENOMEM;
 
+ if (ap->host->n_ports > 1) {
+  pp->irq_desc = devm_kzalloc(dev, 8, GFP_KERNEL);
+  if (!pp->irq_desc) {
+   devm_kfree(dev, pp);
+   return -ENOMEM;
+  }
+  snprintf(pp->irq_desc, 8,
+    "%s%d", dev_driver_string(dev), ap->port_no);
+ }
+
  /* check FBS capability */
  if ((hpriv->cap & HOST_CAP_FBS) && sata_pmp_supported(ap)) {
   void __iomem *port_mmio = ahci_port_base(ap);
-- 
1.7.1